### PR TITLE
Update bond_core to modern cmake.

### PR DESCRIPTION
**Pull request by clalancette**
In particular, make sure it exports targets so downstream users can use those.

While we are here, migrate away from ament_target_dependencies, update to C++17, and move the includes down one directory level (for better workspace overlay support).

### CODE REVIEW

**Review by mjcarroll — APPROVED** [review]

File: smclib/CMakeLists.txt [hunk 1]
```diff
@@ -36,8 +37,8 @@ endif()
 
 ament_python_install_package(${PROJECT_NAME})
 
-ament_export_dependencies(ament_cmake)
-ament_export_include_directories(include)
+ament_export_include_directories(include/${PROJECT_NAME})
```
Comment (Ryanf55) [inline smclib/CMakeLists.txt:40]: If you export targets, what purpose is this line?

Comment (clalancette) [reply smclib/CMakeLists.txt:40]: It's to support downstream packages that haven't yet switched over to targets.  It's a nice way to make sure we can gradually enable targets without forcing all downstream packages to switch at the same time.

**Comment by clalancette** [comment]
This seems to work in my testing locally, both with Humble and with Rolling.  So going ahead and merging this one, thanks for the reviews.

**Comment by felixf4xu** [comment]
as a general rule, should any project with `ament_target_dependencies` be updated?

**Comment by clalancette** [comment]
> as a general rule, should any project with `ament_target_dependencies` be updated?

It's a good question.

The ROS 2 core, as of Jazzy, has been (almost) completely converted to use `target_link_libraries`. But we haven't made any communication about this, sort of on purpose.  There is one major problem, in that you can't currently use a single target for message packages (e.g. `target_link_libraries(lib PRIVATE sensor_msgs::sensor_msgs)` doesn't work).  Instead, you have to do `target_link_libraries(lib PRIVATE ${sensor_msgs_TARGETS})`, which ensures you pull in all of targets necessary.  We need to fix that before we make a general announcement about, and probably write some documentation on how to convert away from `ament_target_dependencies`.

That said, if you are OK with that workaround, it is not a bad idea to get started on the conversion, since we eventually do want to deprecate `ament_target_dependencies`.

**Comment by wep21** [comment]
@clalancette Will you also backport this change into jazzy?

**Comment by clalancette** [comment]
As it stands, Rolling and Jazzy use the same development branch, so I just did a release today of `bond_core` into Jazzy as well.

**Pull request merged**